src/coloring.rs

```diff
@@ -59,3 +67,209 @@ pub fn graph_greedy_color(py: Python, graph: &graph::PyGraph) -> PyResult<PyObje
     }
     Ok(out_dict.into())
 }
+
+struct MisraGriesAlgorithm<'a> {
+    graph: &'a StablePyGraph<Undirected>,
+    colors: DictMap<EdgeIndex, usize>,
+}
+
+impl<'a> MisraGriesAlgorithm<'a> {
+    pub fn new(graph: &'a StablePyGraph<Undirected>) -> Self {
+        let colors: DictMap<EdgeIndex, usize> = DictMap::new();
+        MisraGriesAlgorithm { graph, colors }
+    }
+
+    // Computes colors used at node u
+    fn get_used_colors(&self, u: NodeIndex) -> Vec<usize> {
+        let mut used_colors: Vec<usize> = Vec::new();
+        for edge in self.graph.edges(u) {
+            if let Some(color) = self.colors.get(&edge.id()) {
+                used_colors.push(*color);
+            }
+        }
+        used_colors
```

Comment:
You can actually do this with a single iterator which will be more efficient (mostly because the compiler has more context around the size of things so it can handle allocations better):

```suggestion
        self.graph.edges(u)
            .filter_map(|edge| {
                self.colors.get(&edge.id())
            })
            .copied()
            .collect()
```

(I didn't test this locally so it might not compile, I was mostly guessing from memory about the `copied()` because you'll have a `&usize` for color.

Reply:
Nice, `filter_map` is indeed a useful thing to know about. Yes, I do apparently need `copied()` here (otherwise rust indeed complains about the mismatch between `size` and `&size`).